[HLSL] Parameter modifier parsing and AST

llvm-beanz opened the pull request:
This change implements parsing for HLSL's parameter modifier keywords `in`, `out` and `inout`. Because HLSL doesn't support references or pointers, these keywords are used to allow parameters to be passed in and out of functions.

This change only implements the parsing and AST support. In the HLSL ASTs we represent `out` and `inout` parameters as references, and we implement the semantics of by-value passing during IR generation.

In HLSL parameters marked `out` and `inout` are ambiguous in function declarations, and `in`, `out` and `inout` may be ambiguous at call sites.

This means a function may be defined as `fn(in T)` and `fn(inout T)` or `fn(out T)`, but not `fn(inout T)` and `fn(out T)`. If a funciton `fn` is declared with `in` and `inout` or `out` arguments, the call will be ambiguous the same as a C++ call would be ambiguous given declarations `fn(T)` and `fn(T&)`.

Fixes #59849

llvm-beanz committed:
[HLSL] Parameter modifier parsing and AST

This change implements parsing for HLSL's parameter modifier keywords
`in`, `out` and `inout`. Because HLSL doesn't support references or
pointers, these keywords are used to allow parameters to be passed in
and out of functions.

This change only implements the parsing and AST support. In the HLSL
ASTs we represent `out` and `inout` parameters as references, and we
implement the semantics of by-value passing during IR generation.

In HLSL parameters marked `out` and `inout` are ambiguous in function
declarations, and `in`, `out` and `inout` may be ambiguous at call
sites.

This means a function may be defined as `fn(in T)` and `fn(inout T)` or
`fn(out T)`, but not `fn(inout T)` and `fn(out T)`. If a funciton `fn`
is declared with `in` and `inout` or `out` arguments, the call will be
ambiguous the same as a C++ call would be ambiguous given declarations
`fn(T)` and `fn(T&)`.

python3kgae reviewed (approved):
Maybe a verify test for use rvalue for inout parameter.

bogner commented on clang/lib/Parse/ParseDecl.cpp:
```diff
@@ -5559,7 +5562,6 @@ bool Parser::isTypeSpecifierQualifier() {
   case tok::kw___write_only:
   case tok::kw___funcref:
-  case tok::kw_groupshared:
     return true;
```
This is technically an unrelated bugfix. Not sure it's worth committing it separately, but maybe we should add a test that these keywords are not accepted in C++ mode similar to how we test some of the opencl keywords (see `clang/test/SemaOpenCLCXX/private-access-specifier.cpp`)

bogner commented on clang/test/SemaHLSL/parameter_modifiers.hlsl:
```diff
@@ -0,0 +1,64 @@
+void fn8(inout float f) {}
+
+// These two declare two different functions (although calling them will be
+// ambiguous). This is equivalent to declaring a functiion that takes a
```
Typo. s/functiion/function/

llvm-beanz committed:
Extend testing and handle template cases

Template instantiations for this attribute are a bit odd because HLSL
doesn't allow the construction of reference types from source spellings.

The way that we need to handle this is by not applying the type
modification to template dependent types. Then we can apply the
modification during instantation when the attribute is applied to the
instantiated template type.

This update also includes additional testing to cover parameter
modifier usage in C++. I've also added an additional test case that
verifies errors are produced when trying to use the groupshared keyword
in C++ (a fix of opportunity included in my initial change).

llvm-beanz committed:
[NFC] Fix typo in comment

llvm-beanz committed:
[NFC] Fix formatting
